### PR TITLE
[FIRRTL] Narrow And, Or, and Xor

### PR DESCRIPTION
**Pull request by darthscsi**
Narrow And, Or, and Xor when possible.

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td [hunk 1]
```diff
@@ -272,6 +272,23 @@ def AndOfSelf : Pat <
...
   [(KnownWidth $x)]>;
 
+def TypeWidthAdjust32 : NativeCodeCall<
+  "$_builder.getI32IntegerAttr($0.getType().cast<FIRRTLBaseType>().getBitWidthOrSentinel() - $1.getType().cast<FIRRTLBaseType>().getBitWidthOrSentinel())">;
+
+/// and(x, pad(y, n)) -> pad(and(tail(x), y), n), x is unsigned
+/// And takes an unsigned, so pad must produce an unsigned, which means it's input is unsigned
+def AndOfPadRhs : Pat <
+  (AndPrimOp:$old $x, (PadPrimOp:$pad $y, $n)),
+  (MoveNameHint $old, (PadPrimOp (AndPrimOp (TailPrimOp $x, (TypeWidthAdjust32 $x, $y)), $y), $n)),
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:282]: I think this is correct, even when 'n' is < width of y.  Nice!

File: include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td [hunk 2]
```diff
@@ -272,6 +272,23 @@ def AndOfSelf : Pat <
   (MoveNameHint $old, (AsUIntPrimOp $x)),
   [(KnownWidth $x)]>;
 
+def TypeWidthAdjust32 : NativeCodeCall<
+  "$_builder.getI32IntegerAttr($0.getType().cast<FIRRTLBaseType>().getBitWidthOrSentinel() - $1.getType().cast<FIRRTLBaseType>().getBitWidthOrSentinel())">;
+
+/// and(x, pad(y, n)) -> pad(and(tail(x), y), n), x is unsigned
+/// And takes an unsigned, so pad must produce an unsigned, which means it's input is unsigned
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:279]: And can go from sint to uint apparently?

```firrtl
circuit AndPad:
  module AndPad:
    input x : SInt<3>
    input y : SInt<3>
    output out : UInt

    out <= and(x, pad(y, 4))
```
(which causes verifier failure with this PR)

So I think need to add constraint that x is unsigned (and rest will be per the chain described in comment + existing EqualTypes constraint).

**Review by dtzSiFive — CHANGES_REQUESTED** [review]
I think this looks good, needs another constraint re:signed (see comment).

Is the idea is to hoist pad's out from leaves of expressions?

File: include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td [hunk 3]
```diff
@@ -272,6 +272,23 @@ def AndOfSelf : Pat <
...
+def AndOfPadRhs : Pat <
+  (AndPrimOp:$old $x, (PadPrimOp:$pad $y, $n)),
+  (MoveNameHint $old, (PadPrimOp (AndPrimOp (TailPrimOp $x, (TypeWidthAdjust32 $x, $y)), $y), $n)),
+  [(KnownWidth $x), (EqualTypes $x, $pad)]>;
+
+/// and(x, pad(y, n)) -> pad(and(tail(x), y), n), x is unsigned
```
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:285]: Grep'ing around turned up: "either" in patterns allows specifying order of the next two is interchangeable? Unfortunately not sure can use in result pattern but for 'and' it's probably okay to change this (at least semantically).

Haven't found documentation beyond the code for it yet, FWIW.  https://github.com/llvm/llvm-project/blob/08622314d2a23101536ca744f0092eaec63e14e8/mlir/include/mlir/IR/PatternBase.td#L199 mentions it, and it's used in a test looks like.  Just sharing :+1: .

[the remark this answers is not in the document]
Comment (darthscsi) [reply include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:285]: I've run into that, but didn't see it used anywhere.

[the remark this answers is not in the document]
Comment (darthscsi) [reply include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td]: but it seems to work well enough.  With it preserved order, but oh well.

File: include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td [hunk 4]
```diff
@@ -245,13 +245,13 @@ def SubOfSelf : Pat <
 // sub((pad a, n), b) -> pad(sub(a, b), n)
 def SubOfPadL : Pat <
   (SubPrimOp:$old (PadPrimOp $a, $m), $b),
-  (PadPrimOp (MoveNameHint $old, (SubPrimOp $a, $b)), (TypeWidth32 $old)),
+  (MoveNameHint $old, (PadPrimOp (SubPrimOp $a, $b), (TypeWidth32 $old))),
```
[the remark this answers is not in the document]
Comment (darthscsi) [reply include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:242]: I was rather torn actually.  The op the user thinks about just got narrowed, so it makes sense the name stays.  The value the op produces is a new op, so it makes sense the name moves.

What is named?  The value or the op?

**Pull request merged**